Query mappings fail if the URL has multiple values for a parameter name

At present mapping a request fails if:
1. There are any query parameter mappings;
2. The request contains any parameters that are present more than once.

Note that the parameter with multiple values need not be mapped for the error to occur.

To resolve this:
1. Mapping shouldn't crash under these circumstances;
2. We should define what happens if a field is mapped that has multiple values.

For reference, the stack trace currently caused by this error looks like this:

```
java.util.concurrent.CompletionException: java.lang.IllegalStateException: Duplicate key aqa
        at java.util.concurrent.CompletableFuture.internalComplete(CompletableFuture.java:205) ~[na:1.8.0_20]
        at java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:459) [na:1.8.0_20]
        at java.util.concurrent.CompletableFuture$Async.run(CompletableFuture.java:428) [na:1.8.0_20]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_20]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_20]
        at java.lang.Thread.run(Thread.java:745) [na:1.8.0_20]
Caused by: java.lang.IllegalStateException: Duplicate key aqa
        at java.util.stream.Collectors.lambda$throwingMerger$135(Collectors.java:133) ~[na:1.8.0_20]
        at java.util.stream.Collectors$$Lambda$34/2081964843.apply(Unknown Source) ~[na:na]
        at java.util.HashMap.merge(HashMap.java:1245) ~[na:1.8.0_20]
        at java.util.stream.Collectors.lambda$toMap$193(Collectors.java:1320) ~[na:1.8.0_20]
        at java.util.stream.Collectors$$Lambda$36/1086323658.accept(Unknown Source) ~[na:na]
        at java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169) ~[na:1.8.0_20]
        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1374) ~[na:1.8.0_20]
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:512) ~[na:1.8.0_20]
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:502) ~[na:1.8.0_20]
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708) ~[na:1.8.0_20]
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[na:1.8.0_20]
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499) ~[na:1.8.0_20]
        at io.divolte.server.RecordMapper$Context.lambda$queryStringParameterMap$123(RecordMapper.java:548) ~[divolte-collector-0.1-0.1-SNAPSHOT.jar:na]
        at io.divolte.server.RecordMapper$Context$$Lambda$221/409809083.apply(Unknown Source) ~[na:na]
        at java.util.HashMap.computeIfAbsent(HashMap.java:1118) ~[na:1.8.0_20]
        at io.divolte.server.RecordMapper$Context.queryStringParameterMap(RecordMapper.java:533) ~[divolte-collector-0.1-0.1-SNAPSHOT.jar:na]
        at io.divolte.server.RecordMapper.lambda$null$34(RecordMapper.java:227) ~[divolte-collector-0.1-0.1-SNAPSHOT.jar:na]
        at io.divolte.server.RecordMapper$$Lambda$220/1426299870.apply(Unknown Source) ~[na:na]
        at java.util.Optional.map(Optional.java:215) ~[na:1.8.0_20]
        at io.divolte.server.RecordMapper.lambda$queryParamFieldSupplier$35(RecordMapper.java:227) ~[divolte-collector-0.1-0.1-SNAPSHOT.jar:na]
        at io.divolte.server.RecordMapper$$Lambda$117/1485485458.get(Unknown Source) ~[na:na]
        at io.divolte.server.RecordMapper.lambda$fieldSetterFromConfig$18(RecordMapper.java:128) ~[divolte-collector-0.1-0.1-SNAPSHOT.jar:na]
        at io.divolte.server.RecordMapper$$Lambda$41/553821397.setFields(Unknown Source) ~[na:na]
        at io.divolte.server.RecordMapper.lambda$newRecordFromExchange$118(RecordMapper.java:454) ~[divolte-collector-0.1-0.1-SNAPSHOT.jar:na]
        at io.divolte.server.RecordMapper$$Lambda$146/91823676.accept(Unknown Source) ~[na:na]
        at java.util.ArrayList.forEach(ArrayList.java:1249) ~[na:1.8.0_20]
        at io.divolte.server.RecordMapper.newRecordFromExchange(RecordMapper.java:454) ~[divolte-collector-0.1-0.1-SNAPSHOT.jar:na]
        at io.divolte.server.IncomingRequestProcessor.process(IncomingRequestProcessor.java:72) ~[divolte-collector-0.1-0.1-SNAPSHOT.jar:na]
        at io.divolte.server.IncomingRequestProcessor.process(IncomingRequestProcessor.java:28) ~[divolte-collector-0.1-0.1-SNAPSHOT.jar:na]
        at io.divolte.server.processing.ItemProcessor.process(ItemProcessor.java:16) ~[divolte-collector-0.1-0.1-SNAPSHOT.jar:na]
        at io.divolte.server.processing.ProcessingPool.lambda$microBatchingQueueDrainerWithHeartBeat$5(ProcessingPool.java:132) ~[divolte-collector-0.1-0.1-SNAPSHOT.jar:na]
        at io.divolte.server.processing.ProcessingPool$$Lambda$10/2081652693.run(Unknown Source) ~[na:na]
        at java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:454) [na:1.8.0_20]
        ... 4 common frames omitted
```

--- a/src/main/java/io/divolte/server/RecordMapper.java
+++ b/src/main/java/io/divolte/server/RecordMapper.java
@@ -46,6 +46,7 @@ import org.apache.http.client.utils.URLEncodedUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import com.google.common.base.Strings;
 import com.google.common.cache.CacheBuilder;
 import com.google.common.cache.CacheLoader;
 import com.google.common.cache.LoadingCache;
@@ -542,10 +543,13 @@ final class RecordMapper {
                 if (rawQuery == null) {
                     return Collections.emptyMap();
                 } else {
-                    return URLEncodedUtils
-                            .parse(rawQuery, StandardCharsets.UTF_8)
-                            .stream()
-                            .collect(Collectors.toMap(NameValuePair::getName, NameValuePair::getValue));
+                    final List<NameValuePair> parsedParameters =
+                            URLEncodedUtils.parse(rawQuery, StandardCharsets.UTF_8);
+                    return parsedParameters.stream()
+                            .collect(Collectors.toMap(NameValuePair::getName,
+                                     (p) -> Strings.nullToEmpty(p.getValue()),
+                                     (v1, v2) -> v1,
+                                     () -> Maps.newHashMapWithExpectedSize(parsedParameters.size())));
                 }
             });
         }

--- a/src/test/java/io/divolte/server/RecordMapperTest.java
+++ b/src/test/java/io/divolte/server/RecordMapperTest.java
@@ -256,6 +256,66 @@ public class RecordMapperTest {
     }
 
     @Test
+    public void shouldSetFieldWithEmptyQueryParameterValue() throws IOException, UnirestException {
+        Schema schema = schemaFromClassPath("/TestRecord.avsc");
+        Config config = ConfigFactory.load("schema-test-queryparam");
+        RecordMapper maker = new RecordMapper(schema, config, ConfigFactory.load(), Optional.empty());
+
+        setupExchange(
+                "Divolte/Test",
+                "l=http://ci-website-elb-1121474138.eu-west-1.elb.amazonaws.com/search?q=",
+                "r=https://www.example.com/about.html");
+        GenericRecord record = maker.newRecordFromExchange(theExchange);
+
+        assertEquals("", record.get("queryparam").toString());
+    }
+
+    @Test
+    public void shouldSetFieldWithMissingQueryParameterValue() throws IOException, UnirestException {
+        Schema schema = schemaFromClassPath("/TestRecord.avsc");
+        Config config = ConfigFactory.load("schema-test-queryparam");
+        RecordMapper maker = new RecordMapper(schema, config, ConfigFactory.load(), Optional.empty());
+
+        setupExchange(
+                "Divolte/Test",
+                "l=http://ci-website-elb-1121474138.eu-west-1.elb.amazonaws.com/search?q",
+                "r=https://www.example.com/about.html");
+        GenericRecord record = maker.newRecordFromExchange(theExchange);
+
+        assertEquals("", record.get("queryparam").toString());
+    }
+
+    @Test
+    public void shouldSetFieldWithFirstQueryParameterValue() throws IOException, UnirestException {
+        Schema schema = schemaFromClassPath("/TestRecord.avsc");
+        Config config = ConfigFactory.load("schema-test-queryparam");
+        RecordMapper maker = new RecordMapper(schema, config, ConfigFactory.load(), Optional.empty());
+
+        setupExchange(
+                "Divolte/Test",
+                "l=http://ci-website-elb-1121474138.eu-west-1.elb.amazonaws.com/search?q=first+param&q=second+param",
+                "r=https://www.example.com/about.html");
+        GenericRecord record = maker.newRecordFromExchange(theExchange);
+
+        assertEquals("first param", record.get("queryparam").toString());
+    }
+
+    @Test
+    public void shouldNotFailIfQueryParameterWithMissingNameExists() throws IOException, UnirestException {
+        Schema schema = schemaFromClassPath("/TestRecord.avsc");
+        Config config = ConfigFactory.load("schema-test-queryparam");
+        RecordMapper maker = new RecordMapper(schema, config, ConfigFactory.load(), Optional.empty());
+
+        setupExchange(
+                "Divolte/Test",
+                "l=http://ci-website-elb-1121474138.eu-west-1.elb.amazonaws.com/search?q=a+param&=no+name",
+                "r=https://www.example.com/about.html");
+        GenericRecord record = maker.newRecordFromExchange(theExchange);
+
+        assertEquals("a param", record.get("queryparam").toString());
+    }
+
+    @Test
     public void shouldSetFieldFromCustomEventParameter() throws IOException, UnirestException {
         final Schema schema = schemaFromClassPath("/TestRecord.avsc");
         final Config config = ConfigFactory.load("schema-test-customparameter");

